File: leetcode/python/P0097_InterleavingString.py
# https://leetcode.com/problems/interleaving-string/


class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        # keep one "pointer" for each of the strings;
        # if the letter in s3 corresponds to one of the 2 strings
        # advance
        # it matters from which string we advance:
        # e.g. s1 = abbc; s2 = addc; s3 addabbcc
        #   so 2 pointers would not work

        # if one of the strings is empty, pick everything from
        # the other string;

        # interleaved correctly up to "index" [i, j] if s3[i + j] == s1[i]
        # and match(i-1, j) or s3[i + j] == s2[j] and match(i, j -1)

        if len(s1) + len(s2) != len(s3):
            return False

        dp = [[False for _ in range(len(s1) + 1)] for _ in range(len(s2) + 1)]

        dp[0][0] = True

        # pay attention:
        for idx, ch in enumerate(s1):
            dp[0][idx + 1] = dp[0][idx] and s1[idx] == s3[idx]

        for idx, ch in enumerate(s2):
            dp[idx + 1][0] = dp[idx][0] and s2[idx] == s3[idx]

        for idx1 in range(1, len(s1) + 1):
            for idx2 in range(1, len(s2) + 1):
                dp[idx2][idx1] = (
                    dp[idx2][idx1 - 1] and s3[idx1 + idx2 - 1] == s1[idx1 - 1]
                ) or (dp[idx2 - 1][idx1] and s3[idx1 + idx2 - 1] == s2[idx2 - 1])

        print(dp)

        return dp[len(s2)][len(s1)]

File: leetcode/cpp/P0097_InterleavingString.cpp
// https://leetcode.com/problems/interleaving-string/

#include <iostream>
#include <string>
#include <vector>

class Solution {
 public:
  bool isInterleave(const std::string& s1, const std::string& s2,
                    const std::string& s3) {
    if (s1.length() + s2.length() != s3.length()) {
      return false;
    }

    std::vector<std::vector<int>> dp(s1.length() + 1,
                                     std::vector<int>(s2.length() + 1, 0));

    for (int ii = 0; ii <= s1.length(); ++ii) {
      for (int jj = 0; jj <= s2.length(); ++jj) {
        if (ii == 0 && jj == 0) {
          dp[ii][jj] = 1;
        } else if (ii == 0) {
          dp[ii][jj] = dp[ii][jj - 1] && s2[jj - 1] == s3[ii + jj - 1];
        } else if (jj == 0) {
          dp[ii][jj] = dp[ii - 1][jj] && s1[ii - 1] == s3[ii + jj - 1];
        } else {
          dp[ii][jj] = (dp[ii - 1][jj] && s1[ii - 1] == s3[ii + jj - 1]) ||
                       (dp[ii][jj - 1] && s2[jj - 1] == s3[ii + jj - 1]);
        }
        // cout << dp[ii][jj] << " ";
      }
      // cout << "\n";
    }
    return dp[s1.length()][s2.length()] == 1;
  }
};

int main(int argc, char** argv) {
  Solution s;
  std::cout << s.isInterleave("aabcc", "dbbca", "aadbbcbcac")
            << " Expected: 1.\n";
  std::cout << s.isInterleave("aabcc", "dbbca", "aadbbbaccc")
            << " Expected: 0.\n";
  return 0;
}